Make simplified download button (Fixes #8860)

alexgibson opened the pull request:
## Description
- Adds a new simplified download button helper.
- Implements the simple button on the following pages / areas:
  - Download button in the global navigation.
  - Download buttons on http://localhost:8000/en-US/firefox/new/
  - Download buttons on home pages:
    - http://localhost:8000/en-US/
    - http://localhost:8000/de/
    - http://localhost:8000/fr/
    - http://localhost:8000/es-ES/

## Issue / Bugzilla link
#8860

## Testing
Demo: https://www-demo2.allizom.org/en-US/firefox/new/

- [x] Clicking download on a desktop browser navigates to /thanks and initiates a download as expected.
- [x] Clicking download in an old IE browser initiates a download via a popup, before navigating to /thanks as expected.
- [x] Clicking download on Android / iOS navigated to /thanks before redirecting to the app/play store.

Successful test run: https://gitlab.com/mozmeao/www-config/-/pipelines/188897807

alexgibson commented:
This needs to be reviewed and merged after https://github.com/mozilla/bedrock/pull/9274

alexgibson commented:
This is ready for code review, but do-not-merge pending GA testing / verification.

stephaniehobson commented on bedrock/firefox/templatetags/helpers.py:
```diff
@@ -173,6 +173,59 @@ def download_firefox(ctx, channel='release', platform='all',
+
+    channel = 'release'
+    locale = locale or get_locale(ctx['request'])
+    funnelcake_id = ctx.get('funnelcake_id', False)
```
How does this work? If a `funnelcake_id` is defined somewhere on the page which calls the button this picks that up?

alexgibson replied:
If there's a funnelcake param in the url (e.g. ?f=123) then that will get passed to the /thanks page (e.g. /download/thanks/?f=123). I'll add a comment to document what it does.

stephaniehobson replied:
So we'd still have to use the other button for a campaign page that has a funnel cake? (That's okay, I just want to understand).

alexgibson replied:
No, funnel cakes should work for both button helpers 🙂

stephaniehobson commented on bedrock/firefox/templatetags/helpers.py:
```diff
@@ -173,6 +173,59 @@ def download_firefox(ctx, channel='release', platform='all',
+        channel, version, 'win', locale,
+        force_direct=True,
+        force_full_installer=False,
+        force_funnelcake=False,
```
The original `download_firefox` has some documentation about what these mean. If this function will eventually replace that one it would be nice to preserve the documentation.

alexgibson replied:
I don't think this will replace the full download button entirely as it's needed to download pre-release products still, but good call to add some more docs 👍

stephaniehobson reviewed (commented):
I haven't started looking at the JS yet but here are a few comments from my initial pass.

stephaniehobson commented on media/js/base/mozilla-utils.js:
```diff
@@ -21,6 +21,63 @@ if (typeof window.Mozilla === 'undefined') {
+            if (data && data.os && data.name && data.version) {
+                downloadButton[i].setAttribute('data-display-os', data.os);
+                downloadButton[i].setAttribute('data-display-name', data.name);
+                downloadButton[i].setAttribute('data-download-version', data.version);
```
There isn't a 1:1 mapping for the version defined by the old button to the version defined in the JS here. (For example win64, win-msi, and standard will be reported as something else) Is that okay?

alexgibson replied:
Win64 downloads have never been enabled via the dl button (since the stub installer handles detection), and I don't think we'd ever serve msi installers via /thanks. So I think this is OK.

stephaniehobson replied:
There's not a lot of them but win64, win64-msi are currently being [reported in GA](https://analytics.google.com/analytics/web/#/report/content-event-events/a36116321w64087799p65789850/explorer-table.secSegmentId=analytics.customDimension27&explorer-table.plotKeys=%5B%5D&explorer-table.rowCount=250&explorer-table.filter=win). I don't see standard though, even though it is documented in the read the docs.

alexgibson replied:
There’s a modal on /windows that enables downloading specific builds. On /browsers/windows-64-bit/ my guess is they are following the link to the /all page.

I don’t think any of these events are coming from download buttons since those links have never been enabled to appear.

stephaniehobson replied:
If you[ filter to /firefox/new](https://analytics.google.com/analytics/web/#/report/content-event-events/a36116321w64087799p65789850/explorer-table.plotKeys=%5B%5D&explorer-table.secSegmentId=analytics.customDimension27&_r.drilldown=analytics.eventCategory:~2Ffirefox~2Fnew~2F%20Interactions,analytics.eventAction:Firefox%20Download/) you can see they are being triggered on /new 🤷‍♀️

stephaniehobson replied:
Is that the modal there too?

stephaniehobson replied:
Hm, also seeing them on [/firefox](https://analytics.google.com/analytics/web/#/report/content-event-events/a36116321w64087799p65789850/_r.drilldown=analytics.eventCategory:~2Ffirefox~2F%20Interactions&explorer-table.secSegmentId=analytics.customDimension27&explorer-table.plotKeys=%5B%5D/). No modal there.

alexgibson replied:
The older version of /new has the modal yes. Are these session based events? That may explain things. Still think it this is likely a red herring.

stephaniehobson commented:
I can't wait to see the impact implementing this in the nav bar has on performance! 🔥

alexgibson commented on bedrock/firefox/templates/firefox/includes/download-button-thanks.html:
```diff
@@ -0,0 +1,19 @@
+<div class="mzp-c-button-download-container c-button-download-thanks">
```
You can already add a custom button class and ID to the link itself. I don't think it makes sense to have another set for the wrapper, so if it's more useful I'll move them both to the top HTML element?

alexgibson replied:
Ok, we need to have a custom button class on the link so we can do things like `mzp-t-secondary`. It might be potentially confusing to have a different param and class for the wrapper, so what I propose is to make the ID belong to the top level element. Does that work?

stephaniehobson replied:
ID on top-level +1.

I think that's best until we can have parent selectors in CSS.

alexgibson commented on bedrock/firefox/templates/firefox/includes/download-button-thanks.html:
```diff
@@ -0,0 +1,19 @@
+<div class="mzp-c-button-download-container c-button-download-thanks">
+  <a href="{{ transition_url }}"
+     id="{{ id }}"
+     class="mzp-c-button mzp-t-product {% if button_class %}{{ button_class }}{% else %}mzp-t-xl{% endif %}"
```
It's not strictly needed, but I think having a singular identifiable class for the link (rather than needing `.c-button-download-thanks > a` makes sense. For compatibility purposes `download-link` wfm 👍

alexgibson commented on bedrock/firefox/templates/firefox/includes/download-button-thanks.html:
```diff
@@ -0,0 +1,19 @@
+      {{ ftl('download-button-download-firefox') }}
+    {% endif %}
+  </a>
+  <small class="mzp-c-button-download-privacy-link">
```
It looks like the new button was used on 2/5 of those .scss files. I've updated the ones that were effected. nice catch 👍

alexgibson commented:
Thanks @stephaniehobson - I pushed a commit that addresses some feedback. I also added some docs for the download button to try and help people understand which one to use and when.

alexgibson commented:
@pmac r on the python bits here? I'm mostly wondering if I've overlooked anything that should still be included in the newer helper.

pmac commented on bedrock/firefox/templatetags/helpers.py:
```diff
@@ -173,6 +173,60 @@ def download_firefox(ctx, channel='release', platform='all',
+    funnelcake_id = ctx.get('funnelcake_id', False)
+    dom_id = dom_id or 'download-button-thanks'
+    transition_url = '/firefox/download/thanks/'
+    version = firefox_desktop.latest_version(channel)
```
This is not something we should fix in this PR, but the `version` parameter to the `get_download_url` isn't even used in the function anymore. It's a vestige of the old way of generating direct download links before we only used the `*-latest` bouncer aliases. This doesn't hurt anything, but you could also just pass in `None` for `version` and it would work the same.

pmac commented on bedrock/firefox/tests/test_helpers.py:
```diff
@@ -310,6 +310,59 @@ def test_ios(self):
+        link = pq(links)
+        href = link.attr('href')
+
+        assert href == ('/en-US/firefox/download/thanks/')
```
style nit: superfluous parens

pmac commented on bedrock/firefox/tests/test_helpers.py:
```diff
@@ -310,6 +310,59 @@ def test_ios(self):
+        link = pq(links)
+        href = link.attr('href')
+
+        assert href == ('/firefox/download/thanks/')
```
style nit: superfluous parens

pmac reviewed (approved):
Looks good! Just a couple of small notes, but 👍 from me.

stephaniehobson commented on bedrock/firefox/templates/firefox/includes/download-button-thanks.html:
```diff
@@ -0,0 +1,18 @@
+  <a href="{{ transition_url }}"
+     class="download-link mzp-c-button mzp-t-product {% if button_class %}{{ button_class }}{% else %}mzp-t-xl{% endif %}"
+     data-direct-link="{{ download_link_direct }}"
+     data-link-type="download"
```
There's a change in the GA reporting. This is showing up with the Event Label "Download for unavailable". I think it is because `data-download-os` is defined on the old button but not this one.

stephaniehobson replied:
Otherwise GA reporting looks to be the same in Chrome/MacOS:

| Prod     | Demo |
| --- | --- |
| ![Screen Shot 2020-09-15 at 07 22 12](https://user-images.githubusercontent.com/854701/93225064-a9ee3500-f726-11ea-9b7a-d837ce323814.png) | ![Screen Shot 2020-09-15 at 07 24 19](https://user-images.githubusercontent.com/854701/93225083-ad81bc00-f726-11ea-84de-02bf641d5d9a.png) |

alexgibson replied:
Strange, because I'm adding `data-download-os` via JS (since the value is dynamic)

alexgibson replied:
I can see `data-download-os` on the link when viewing source on demo:

<img width="1674" alt="image" src="https://user-images.githubusercontent.com/400117/93226105-de83e000-f76a-11ea-8039-79029a54415b.png">

alexgibson replied:
Oh wait - it's `data-display-os` (face palm). At least that's a sinple fix :)

stephaniehobson replied:
Oh, I was viewing source. Then I don't know what's causing the difference. Are you able to look into it?

alexgibson replied:
Yeah I think I've got it - typo (see comment above ^) :)

stephaniehobson replied:
LOL, I made the same mistake in my comment 🤦‍♀️

alexgibson replied:
Updated data attribute to the correct name

stephaniehobson commented on docs/download-buttons.rst:
```diff
@@ -0,0 +1,41 @@
+.. _download-buttons:
+
+========================
+Firefox Download Buttons
```
📄 💯

alexgibson commented:
Leaving do-not-merge in place until adswerve have taken a look in GA

alexgibson commented:
Ok, adswerve have done some QA and confirmed things look good in GA. I'm going to remove the DNM label from the PR, but we should probably wait until Monday to merge and push to prod (pending final review approval).

stephaniehobson reviewed (approved):
R+ 🦋
